### PR TITLE
Issue 533 update envoy 16319 jwks retries flags

### PR DESCRIPTION
**Pull request by alichnewsky**
update envoy version to one supporting truncated exponential backoff on remote jwks fetches failures.
add new flags to turn on that functionality in esp-v2. 
default : no retries (behaviour unchanged)

### CODE REVIEW

**Comment by alichnewsky** [comment]
this PR addresses issue #533.
The required functionality finally made it into envoy

File: src/go/options/configgenerator.go [hunk 1]
```diff
@@ -108,8 +108,11 @@ type ConfigGeneratorOptions struct {
 	ConnectionBufferLimitBytes    int
 
 	// JwtAuthn related flags
-	DisableJwksAsyncFetch bool
-	JwksCacheDurationInS  int
+	DisableJwksAsyncFetch             bool
```
Comment (nareddyt) [inline src/go/options/configgenerator.go:111]: Diff looks weird here, is there a change in whitespace? Can you run `make format`?

**Review by nareddyt — COMMENTED** [review]
Implementation LGTM. Please add a unit test for the new config generated when `serviceInfo.Options.JwksFetchNumRetries > 0`

It will be very similar to how `disableJwksAsyncFetch ` is unit tested: https://github.com/GoogleCloudPlatform/esp-v2/blob/4c7e67acf43325ce7cbc56bcdea468166b621be2/src/go/configgenerator/filterconfig/filter_gen_jwt_authn_test.go#L120

File: src/go/options/configgenerator.go (hunk 1 above)
Comment (alichnewsky) [reply src/go/options/configgenerator.go:111]: make format generates identical code.
I think the go formatting takes this line and the next four and decides to align based on the widest variable name
`JwksFetchRetryBackOffBaseInterval` three lines below

File: WORKSPACE [hunk 2]
```diff
@@ -39,9 +39,9 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 # 3) Check if envoy_build_config/extensions_build_config.bzl is up-to-date.
 # Try to match it with the one in source/extensions and comment out unneeded extensions.
 
-ENVOY_SHA1 = "9ebffc61cfd2ea11807757b12a4ae7b92ab0dc8c"  # 2021-06-09: top of the tree
+ENVOY_SHA1 = "b62dae29a5dd06b7f689899b26974d9567a98f0e"  # 2021-07-15: top of the tree
```
[the remark this answers is not in the document]
Comment (qiwzhang) [reply WORKSPACE:42]: Should be fine.  We can make this as exception, otherwise, users have to wait for at lease one month for another envoy release.

**Review by qiwzhang — COMMENTED** [review]
Could you add an integration test [here](https://github.com/GoogleCloudPlatform/esp-v2/blob/master/tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go) for the jwks_retries flag?

**Comment by alichnewsky** [comment]
> Could you add an integration test [here](https://github.com/GoogleCloudPlatform/esp-v2/blob/master/tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go) for the jwks_retries flag?

added a new test with last commit.

**Comment by alichnewsky** [comment]
asan test fails consistently. The pod running it seems to disappear exactly after 1h, and test fails with "pod got deleted unexpectedly". Most likely it is a CI infrastructure configuration issue, and I am not quite sure what I can do about it.

File: tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go [hunk 3]
```diff
@@ -160,6 +191,17 @@ func TestAuthJwksAsyncFetch(t *testing.T) {
...
 		},
+		{
+			desc:                "Success, the jwks provider succeeds only once every 5 requests so at least 5 request to the jwks provider will be made",
+			path:                "/auth/info/auth0",
+			apiKey:              "api-key",
+			method:              "GET",
+			jwksFetchNumRetries: 5,
+			token:               testdata.FakeCloudTokenMultiAudiences,
+			initProviderCnt:     &expectedRequestCount{provider, 1},
+			afterProviderCnt:    &expectedRequestCount{provider, 5},
```
Comment (qiwzhang) [inline tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go:202]: Hmm,  for async_fetch,  all 5 retry should happen before listener is activated.  or before Envoy is ready,  it it should be all in "initProviderCnt".  Could you help to debug it to see why that is not the case

You can run the test by
```
go test -v ./tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go --debug_components=envoy --logtostderr &> /tmp/mmm.log
```

Comment (alichnewsky) [reply tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go:202]: the test done on initProviderCount was "I received at least one request" not "I received exactly one request".
If I log the exact number of request obtained it is indeed 5. And I should test for that.

with a provider in this test 
```
 {
  key: "google_jwt"
  value {
    issuer: "api-proxy-testing@cloud.goog"
    audiences: "https://echo-api.endpoints.cloudesf-testing.cloud.goog"
    remote_jwks {
      http_uri {
        uri: "http://127.0.0.1:20113"
        cluster: "jwt-provider-cluster-127.0.0.1:20113"
        timeout {
          seconds: 30
        }
      }
      cache_duration {
        seconds: 300
      }
      async_fetch {
      }
      retry_policy {
        retry_back_off {
          base_interval {
            nanos: 200000000
          }
          max_interval {
            seconds: 32
          }
        }
        num_retries {
          value: 5
        }
      }
    }
    forward: true
    from_headers {
      name: "Authorization"
      value_prefix: "Bearer "
    }
    from_headers {
      name: "X-Goog-Iap-Jwt-Assertion"
    }
    from_params: "access_token"
    forward_payload_header: "X-Endpoint-API-UserInfo"
    payload_in_metadata: "jwt_payloads"
  }
}
```

It wasn't clear to me that the envoy proxy ingress listener would attempt all retries to its provider JwksFetcher during the set up... I thought I could only guarantee one call at minimum.
Then upon success of a JWT-authorized request, I could guarantee all five retries had occurred, the last one returning a proper JWKS, and the JWT being validated...

I'll modify the test's initProviderCnt accordingly.

Comment (qiwzhang) [reply tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go]: I see. If verifyProviderCnt() is checking minimum request count,  change initProviderCnt to 5 is good enough.  Thanks

**Review by qiwzhang — APPROVED** [review]

File: tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go [hunk 4]
```diff
@@ -64,6 +65,16 @@ func TestAuthJwksCache(t *testing.T) {
...
 			wantResp:               `{"aud":["admin.cloud.goog","bookstore_test_client.cloud.goog"],"exp":4698318999,"iat":1544718999,"iss":"api-proxy-testing@cloud.goog","sub":"api-proxy-testing@cloud.goog"}`,
 		},
+		{
+			desc:                   "Success, the customized jwks succeeds once every five requests so 5 request to the jwks provider will be made",
+			path:                   "/auth/info/auth0",
+			apiKey:                 "api-key",
+			method:                 "GET",
+			jwksFetchNumRetries:    5,
+			token:                  testdata.FakeCloudTokenMultiAudiences,
+			wantRequestsToProvider: &expectedRequestCount{provider, 5},
```
Comment (TAOXUY) [inline tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go:75]: Remove as it is not used?

File: tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go [hunk 5]
```diff
@@ -42,6 +42,7 @@ func TestAuthJwksCache(t *testing.T) {
 		token                  string
 		apiKey                 string
 		jwksCacheDurationInS   int
+		jwksFetchNumRetries    int
 		wantRequestsToProvider *expectedRequestCount
```
Comment (TAOXUY) [inline tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go:46]: Remove as it is not used?

**Review by TAOXUY — APPROVED** [review]

**Comment by alichnewsky** [comment]
I am at my wits end when it comes to making the [ESPv2-presubmit-asan test run](https://oss-prow.knative.dev/log?container=test&id=1420796097764265984&job=ESPv2-presubmit-asan).
What seemed straight CI infrastructure timeouts now seem to be HTTPS timeout downloading go code dependencies...

```
go: downloading github.com/miekg/dns v1.1.29
go: downloading github.com/gorilla/websocket v1.4.2
tests/endpoints/echo/client/client.go:30:2: github.com/gorilla/websocket@v1.4.2: Get "https://proxy.golang.org/github.com/gorilla/websocket/@v/v1.4.2.zip": net/http: TLS handshake timeout
tests/env/components/dns_resolver.go:23:2: github.com/miekg/dns@v1.1.29: Get "https://proxy.golang.org/github.com/miekg/dns/@v/v1.1.29.zip": net/http: TLS handshake timeout
make: *** [Makefile:53: build-msan] Error 1
{"component":"entrypoint","error":"wrapped process failed: exit status 2","file":"prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2021-07-29T17:54:26Z"}
```

**Comment by qiwzhang** [comment]
Hi @alichnewsky our ESPv2-presubmit-asan may be flaky.  we can ignore its failure.

**Comment by alichnewsky** [comment]
> Hi @alichnewsky our ESPv2-presubmit-asan may be flaky. we can ignore its failure.

Great.
Thank you for shepherding me through this all.

**Pull request merged**